File: ForceConnector/processCustomLabelTranslationUpload.vb
Imports ForceConnector.MiniMETA
Imports System.ComponentModel
Imports System.Windows.Forms

Public Class processCustomLabelTranslationUpload

    Dim numOfCustomLabel As Long

    Dim statusText As String = ""

    Dim excelApp As Excel.Application
    Dim workbook As Excel.Workbook
    Dim worksheet As Excel.Worksheet

    Dim start As Excel.Range

    Dim m_table As Excel.Range
    Dim m_start As Excel.Range
    Dim m_head As Excel.Range
    Dim m_body As Excel.Range

    Dim m_rows As Integer ' row pointer to append new value at bottom of list
    Dim m_metaType As String

    Dim m_langSet As List(Of String) = New List(Of String)

    Private Sub processCustomLabelTranslationUpload_Load(sender As Object, e As EventArgs) Handles Me.Load
        Try
            lblMessage.Font = New Drawing.Font(lblMessage.Font, Drawing.FontStyle.Regular)
            lblMessage.ForeColor = Drawing.Color.Black
            statusText = ""

            If Not setMetaBinding() Then
                statusText = "Session Failed"
                GoTo errors
            End If

            '' these properties should be set to True (at design-time or runtime) before calling the RunWorkerAsync
            '' to ensure that it supports Cancellation and reporting Progress
            bgw.WorkerSupportsCancellation = True
            bgw.WorkerReportsProgress = True

            '' call this method to start your asynchronous Task.
            bgw.RunWorkerAsync()

            GoTo done
        Catch ex As Exception
            statusText = ex.Message
        End Try

errors:
        If statusText <> "" Then
            lblMessage.Font = New Drawing.Font(lblMessage.Font, Drawing.FontStyle.Bold)
            lblMessage.ForeColor = Drawing.Color.Red
            lblMessage.Text = statusText
            btnAction.Text = "Done"
            btnAction.Enabled = True
        End If
done:
    End Sub

    Private Sub btnAction_Click(sender As Object, e As EventArgs) Handles btnAction.Click
        '' to cancel the task, just call the BackgroundWorker.CancelAsync method.
        If btnAction.Text = "Cancel" Then
            bgw.CancelAsync()
        ElseIf btnAction.Text = "Done" Then
            MyBase.Close()
        End If
    End Sub

    Private Sub bgw_DoWork(sender As Object, e As DoWorkEventArgs) Handles bgw.DoWork
        '' The asynchronous task you want to perform goes here
        '' the following is an example of how it typically goes.

        Try
            excelApp = ThisAddIn.excelApp
            workbook = excelApp.ActiveWorkbook
            excelApp.StatusBar = "Upload CustomLabel Translation metadata..."
            setControlText(btnAction, "Wait...")
            bgw.ReportProgress(0, "Please wait for initialize...")

            getTranslations(m_langSet)
            If m_langSet.Count = 0 Then
                statusText = "No translation setting in ORG."
                GoTo errors
            End If

            '' Call initialize process...
            worksheet = workbook.ActiveSheet
            If worksheet.Name = "CustomLabel" Then
                setWorkArea(excelApp, worksheet, m_table, m_head, m_body, m_start, m_rows, m_metaType)

                If Not RegQueryBoolValue(GOAHEAD) Then
                    Dim msg As String = "You are about to UPLOAD: " & CStr(excelApp.Selection.Rows.Count) & " CustomLabel Translation " &
                    " metadata(s) in to Salesforce.com" & vbCrLf
                    If (MsgBox(msg, vbApplicationModal + vbOKCancel + vbExclamation + vbDefaultButton1,
                        "-- Ready to Upload CustomLabel Translation --") = vbCancel) Then
                        statusText = "Upload canceled!"
                        GoTo errors
                    End If
                End If

                setControlText(btnAction, "Cancel")
                setControlStatus(btnAction, True)
            Else
                statusText = "Current worksheet is not CustomLabel!"
                GoTo errors
            End If

            Dim clts As List(Of MiniMETA.CustomLabelTranslation) = New List(Of MiniMETA.CustomLabelTranslation)
            Dim todo As Excel.Range = excelApp.Selection
            If todo.Columns.Count > 1 Then
                statusText = "Only one language can upload!"
                GoTo errors
            End If
            Dim langCell As Excel.Range = excelApp.Intersect(todo.Cells(1, 1).EntireColumn, m_head)
            If langCell Is Nothing Then
                statusText = "Could not find the translatable language"
                GoTo errors
            End If
            If Not m_langSet.Contains(langCell.Value) Then
                statusText = "Select area's language does not supported!"
                GoTo errors
            End If

            Dim rw As Excel.Range
            Dim labelCount As Integer = 1
            Dim numOfLabel As Integer = todo.Rows.Count
            For Each rw In todo.Rows
                Dim nameCell As Excel.Range = excelApp.Intersect(m_body.Cells(1, 1).EntireColumn, rw.EntireRow)
                Dim clt As MiniMETA.CustomLabelTranslation = New MiniMETA.CustomLabelTranslation

                clt.name = nameCell.Value
                clt.label = rw.Value
                clts.Add(clt)

                Dim percent As Integer = CInt(80 * (labelCount / numOfLabel))
                labelCount = labelCount + 1
                bgw.ReportProgress(percent, "Build translations ... " & labelCount.ToString & " / " & numOfLabel.ToString())

                '' check at regular intervals for CancellationPending
                If bgw.CancellationPending Then
                    bgw.ReportProgress(percent, "Cancelling...")
                    Exit For
                End If
            Next

            Dim metas As List(Of MiniMETA.Translations) = New List(Of MiniMETA.Translations)
            Dim meta As MiniMETA.Translations = New MiniMETA.Translations
            meta.fullName = langCell.Value
            meta.customLabels = clts.ToArray
            metas.Add(meta)

            bgw.ReportProgress(90, "Update translation...")

            Dim srs() As MiniMETA.SaveResult = updateMetadata(metas.ToArray)
            excelApp.ScreenUpdating = False

            For Each sr As MiniMETA.SaveResult In srs
                If Not sr.success Then
                    Dim msg As String = ""
                    For Each err As MiniMETA.Error In sr.errors
                        msg = msg & vbCrLf & err.message
                    Next

                    MsgBox(msg, Title:=sr.fullName & " Error")
                End If
            Next sr

            '' any cleanup code go here
            '' ensure that you close all open resources before exitting out of this Method.
            '' try to skip off whatever is not desperately necessary if CancellationPending is True
            srs = Nothing

            excelApp.ScreenUpdating = True
            bgw.ReportProgress(100, "Done.")

            '' set the e.Cancel to True to indicate to the RunWorkerCompleted that you cancelled out
            If bgw.CancellationPending Then
                e.Cancel = True
                bgw.ReportProgress(100, "Cancelled.")
            End If

            GoTo done
        Catch ex As Exception
            statusText = ex.Message
            GoTo errors
        End Try

errors:
        If statusText <> "" Then
            'bgw.CancelAsync()
            e.Cancel = True
        End If
done:
        excelApp.ScreenUpdating = True
    End Sub

    Private Sub bgw_ProgressChanged(sender As Object, e As ProgressChangedEventArgs) Handles bgw.ProgressChanged
        '' This event is fired when you call the ReportProgress method from inside your DoWork.
        '' Any visual indicators about the progress should go here.
        progressDownload.Value = CInt(e.ProgressPercentage)
        lblMessage.Text = e.UserState
    End Sub

    Private Sub bgw_RunWorkerCompleted(sender As Object, e As RunWorkerCompletedEventArgs) Handles bgw.RunWorkerCompleted
        '' This event is fired when your BackgroundWorker exits.
        '' It may have exitted Normally after completing its task, 
        '' or because of Cancellation, or due to any Error.

        If e.Error IsNot Nothing Then
            '' if BackgroundWorker terminated due to error
            'MessageBox.Show(e.Error.Message)
            lblMessage.Font = New Drawing.Font(lblMessage.Font, Drawing.FontStyle.Bold)
            lblMessage.ForeColor = Drawing.Color.Red
            lblMessage.Text = "Error occurred!" & e.Error.Message

        ElseIf e.Cancelled Then
            '' otherwise if it was cancelled
            'MessageBox.Show("Download cancelled!")

            If statusText <> "" Then
                lblMessage.Font = New Drawing.Font(lblMessage.Font, Drawing.FontStyle.Bold)
                lblMessage.ForeColor = Drawing.Color.Red
                lblMessage.Text = statusText
            Else
                lblMessage.Text = "Upload Cancelled!"
            End If
        Else
            '' otherwise it completed normally
            'MessageBox.Show("Download completed!")
            lblMessage.Text = "Upload completed!"
        End If

        btnAction.Text = "Done"
        btnAction.Enabled = True
    End Sub

    '******************************************************
    '* Change controls label in the bgw.DoWork()
    '* ctl -> label, button ...etc
    '******************************************************
    Private Sub setControlText(ByVal ctl As Control, ByVal text As String)
        If ctl.InvokeRequired Then
            ctl.Invoke(New setControlTextInvoker(AddressOf setControlText), ctl, text)
        Else
            ctl.Text = text
        End If
    End Sub
    Private Delegate Sub setControlTextInvoker(ByVal ctl As Control, ByVal text As String)

    Private Sub setControlStatus(ByVal ctl As Control, ByVal enabled As Boolean)
        If ctl.InvokeRequired Then
            ctl.Invoke(New setControlStatusInvoker(AddressOf setControlStatus), ctl, enabled)
        Else
            ctl.Enabled = enabled
        End If
    End Sub
    Private Delegate Sub setControlStatusInvoker(ByVal ctl As Control, ByVal enabled As Boolean)
End Class

File: ForceConnector/processCustomLabelTranslationUpload.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Windows.Forms;
using ForceConnector.MiniMETA;
using Excel = Microsoft.Office.Interop.Excel;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace ForceConnector
{
    public partial class processCustomLabelTranslationUpload
    {
        public processCustomLabelTranslationUpload()
        {
            InitializeComponent();
            _btnAction.Name = "btnAction";
        }

        private long numOfCustomLabel;
        private string statusText = "";
        private Excel.Application excelApp;
        private Excel.Workbook workbook;
        private Excel.Worksheet worksheet;
        private Excel.Range start;
        private Excel.Range m_table;
        private Excel.Range m_start;
        private Excel.Range m_head;
        private Excel.Range m_body;
        private int m_rows; // row pointer to append new value at bottom of list
        private string m_metaType;
        private List<string> m_langSet = new List<string>();

        private void processCustomLabelTranslationUpload_Load(object sender, EventArgs e)
        {
            try
            {
                lblMessage.Font = new System.Drawing.Font(lblMessage.Font, System.Drawing.FontStyle.Regular);
                lblMessage.ForeColor = System.Drawing.Color.Black;
                statusText = "";
                if (!METAAPI.setMetaBinding())
                {
                    statusText = "Session Failed";
                    goto errors;
                }

                // ' these properties should be set to True (at design-time or runtime) before calling the RunWorkerAsync
                // ' to ensure that it supports Cancellation and reporting Progress
                bgw.WorkerSupportsCancellation = true;
                bgw.WorkerReportsProgress = true;

                // ' call this method to start your asynchronous Task.
                bgw.RunWorkerAsync();
                goto done;
            }
            catch (Exception ex)
            {
                statusText = ex.Message;
            }

        errors:
            ;
            if (!string.IsNullOrEmpty(statusText))
            {
                lblMessage.Font = new System.Drawing.Font(lblMessage.Font, System.Drawing.FontStyle.Bold);
                lblMessage.ForeColor = System.Drawing.Color.Red;
                lblMessage.Text = statusText;
                btnAction.Text = "Done";
                btnAction.Enabled = true;
            }

        done:
            ;
        }

        private void btnAction_Click(object sender, EventArgs e)
        {
            // ' to cancel the task, just call the BackgroundWorker.CancelAsync method.
            if (btnAction.Text == "Cancel")
            {
                bgw.CancelAsync();
            }
            else if (btnAction.Text == "Done")
            {
                Close();
            }
        }

        private void bgw_DoWork(object sender, DoWorkEventArgs e)
        {
            // ' The asynchronous task you want to perform goes here
            // ' the following is an example of how it typically goes.

            try
            {
                excelApp = ThisAddIn.excelApp;
                workbook = excelApp.ActiveWorkbook;
                excelApp.StatusBar = "Upload CustomLabel Translation metadata...";
                setControlText(btnAction, "Wait...");
                bgw.ReportProgress(0, "Please wait for initialize...");
                METAAPI.getTranslations(ref m_langSet);
                if (m_langSet.Count == 0)
                {
                    statusText = "No translation setting in ORG.";
                    goto errors;
                }

                // ' Call initialize process...
                worksheet = (Excel.Worksheet)workbook.ActiveSheet;
                if (worksheet.Name == "CustomLabel")
                {
                    METAAPI.setWorkArea(ref excelApp, ref worksheet, ref m_table, ref m_head, ref m_body, ref m_start, ref m_rows, ref m_metaType);
                    if (!RegDB.RegQueryBoolValue(ForceConnector.GOAHEAD))
                    {
                        string msg = "You are about to UPLOAD: " + Conversions.ToString(excelApp.Selection.Rows.Count) + " CustomLabel Translation " + " metadata(s) in to Salesforce.com" + Constants.vbCrLf;
                        if (Interaction.MsgBox(msg, (MsgBoxStyle)((int)Constants.vbApplicationModal + (int)Constants.vbOKCancel + (int)Constants.vbExclamation + (int)Constants.vbDefaultButton1), "-- Ready to Upload CustomLabel Translation --") == Constants.vbCancel)
                        {
                            statusText = "Upload canceled!";
                            goto errors;
                        }
                    }

                    setControlText(btnAction, "Cancel");
                    setControlStatus(btnAction, true);
                }
                else
                {
                    statusText = "Current worksheet is not CustomLabel!";
                    goto errors;
                }

                var clts = new List<CustomLabelTranslation>();
                Excel.Range todo = (Excel.Range)excelApp.Selection;
                if (todo.Columns.Count > 1)
                {
                    statusText = "Only one language can upload!";
                    goto errors;
                }

                var langCell = excelApp.Intersect((Excel.Range)todo.Cells[(object)1, (object)1].EntireColumn, m_head);
                if (langCell is null)
                {
                    statusText = "Could not find the translatable language";
                    goto errors;
                }

                if (!m_langSet.Contains(Conversions.ToString(langCell.get_Value())))
                {
                    statusText = "Select area's language does not supported!";
                    goto errors;
                }

                int labelCount = 1;
                int numOfLabel = todo.Rows.Count;
                foreach (Excel.Range rw in todo.Rows)
                {
                    var nameCell = excelApp.Intersect((Excel.Range)m_body.Cells[(object)1, (object)1].EntireColumn, rw.EntireRow);
                    var clt = new CustomLabelTranslation();
                    clt.name = Conversions.ToString(nameCell.get_Value());
                    clt.label = Conversions.ToString(rw.get_Value());
                    clts.Add(clt);
                    int percent = (int)Math.Round(80d * (labelCount / (double)numOfLabel));
                    labelCount = labelCount + 1;
                    bgw.ReportProgress(percent, "Build translations ... " + labelCount.ToString() + " / " + numOfLabel.ToString());

                    // ' check at regular intervals for CancellationPending
                    if (bgw.CancellationPending)
                    {
                        bgw.ReportProgress(percent, "Cancelling...");
                        break;
                    }
                }

                var metas = new List<Translations>();
                var meta = new Translations();
                meta.fullName = Conversions.ToString(langCell.get_Value());
                meta.customLabels = clts.ToArray();
                metas.Add(meta);
                bgw.ReportProgress(90, "Update translation...");
                var srs = METAAPI.updateMetadata(metas.ToArray());
                excelApp.ScreenUpdating = false;
                foreach (SaveResult sr in srs)
                {
                    if (!sr.success)
                    {
                        string msg = "";
                        foreach (Error err in sr.errors)
                            msg = msg + Constants.vbCrLf + err.message;
                        Interaction.MsgBox(msg, Title: sr.fullName + " Error");
                    }
                }

                // ' any cleanup code go here
                // ' ensure that you close all open resources before exitting out of this Method.
                // ' try to skip off whatever is not desperately necessary if CancellationPending is True
                srs = null;
                excelApp.ScreenUpdating = true;
                bgw.ReportProgress(100, "Done.");

                // ' set the e.Cancel to True to indicate to the RunWorkerCompleted that you cancelled out
                if (bgw.CancellationPending)
                {
                    e.Cancel = true;
                    bgw.ReportProgress(100, "Cancelled.");
                }

                goto done;
            }
            catch (Exception ex)
            {
                statusText = ex.Message;
                goto errors;
            }

        errors:
            ;
            if (!string.IsNullOrEmpty(statusText))
            {
                // bgw.CancelAsync()
                e.Cancel = true;
            }

        done:
            ;
            excelApp.ScreenUpdating = true;
        }

        private void bgw_ProgressChanged(object sender, ProgressChangedEventArgs e)
        {
            // ' This event is fired when you call the ReportProgress method from inside your DoWork.
            // ' Any visual indicators about the progress should go here.
            progressDownload.Value = e.ProgressPercentage;
            lblMessage.Text = Conversions.ToString(e.UserState);
        }

        private void bgw_RunWorkerCompleted(object sender, RunWorkerCompletedEventArgs e)
        {
            // ' This event is fired when your BackgroundWorker exits.
            // ' It may have exitted Normally after completing its task, 
            // ' or because of Cancellation, or due to any Error.

            if (e.Error is object)
            {
                // ' if BackgroundWorker terminated due to error
                // MessageBox.Show(e.Error.Message)
                lblMessage.Font = new System.Drawing.Font(lblMessage.Font, System.Drawing.FontStyle.Bold);
                lblMessage.ForeColor = System.Drawing.Color.Red;
                lblMessage.Text = "Error occurred!" + e.Error.Message;
            }
            else if (e.Cancelled)
            {
                // ' otherwise if it was cancelled
                // MessageBox.Show("Download cancelled!")

                if (!string.IsNullOrEmpty(statusText))
                {
                    lblMessage.Font = new System.Drawing.Font(lblMessage.Font, System.Drawing.FontStyle.Bold);
                    lblMessage.ForeColor = System.Drawing.Color.Red;
                    lblMessage.Text = statusText;
                }
                else
                {
                    lblMessage.Text = "Upload Cancelled!";
                }
            }
            else
            {
                // ' otherwise it completed normally
                // MessageBox.Show("Download completed!")
                lblMessage.Text = "Upload completed!";
            }

            btnAction.Text = "Done";
            btnAction.Enabled = true;
        }

        // ******************************************************
        // * Change controls label in the bgw.DoWork()
        // * ctl -> label, button ...etc
        // ******************************************************
        private void setControlText(Control ctl, string text)
        {
            if (ctl.InvokeRequired)
            {
                ctl.Invoke(new setControlTextInvoker(setControlText), ctl, text);
            }
            else
            {
                ctl.Text = text;
            }
        }

        private delegate void setControlTextInvoker(Control ctl, string text);

        private void setControlStatus(Control ctl, bool enabled)
        {
            if (ctl.InvokeRequired)
            {
                ctl.Invoke(new setControlStatusInvoker(setControlStatus), ctl, enabled);
            }
            else
            {
                ctl.Enabled = enabled;
            }
        }

        private delegate void setControlStatusInvoker(Control ctl, bool enabled);
    }
}